From aedfea4d26cea0d999411745b421dd0faa8506eb Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@linaro.org>
Date: Thu, 30 Mar 2023 18:35:04 +0200
Subject: [PATCH 069/841] feat(stm32mp1): default load optee in DDR

Adds config switch STM32MP1_OPTEE_IN_SYSRAM. When enabled, OP-TEE is
loading in SYSRAM and executes with pager. When disabled, OP-TEE is
executed in DDR. The switch is default disabled.

bl2_el3_plat_prepare_exit() is also updated to flush BL2 data in RAM
when OP-TEE executes in DDR to ensure BL2 low power sequences are
consistent in SYSRAM for OP-TEE to call these resident services.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 59cd273e831513f8e78ffe35c15ab06e46a51da0 feat(stm32mp1): default load optee in DDR]
Change-Id: Id923e02d55506cb9a6554716819d8c90d93e6f55
Signed-off-by: Etienne Carriere <etienne.carriere@linaro.org>
---
 fdts/stm32mp15-fw-config.dtsi       | 10 +++++++++-
 plat/st/stm32mp1/platform.mk        |  7 +++++++
 plat/st/stm32mp1/stm32mp1_fip_def.h |  2 +-
 3 files changed, 17 insertions(+), 2 deletions(-)

diff --git a/fdts/stm32mp15-fw-config.dtsi b/fdts/stm32mp15-fw-config.dtsi
index d9f6beda7..afc083d89 100644
--- a/fdts/stm32mp15-fw-config.dtsi
+++ b/fdts/stm32mp15-fw-config.dtsi
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: (GPL-2.0+ OR BSD-3-Clause)
 /*
- * Copyright (c) 2021-2022, STMicroelectronics - All Rights Reserved
+ * Copyright (c) 2021-2023, STMicroelectronics - All Rights Reserved
  */
 
 #include <common/tbbr/tbbr_img_def.h>
@@ -41,11 +41,19 @@
 		};
 
 #ifdef AARCH32_SP_OPTEE
+#if STM32MP1_OPTEE_IN_SYSRAM
 		tos_fw {
 			load-address = <0x0 STM32MP_OPTEE_BASE>;
 			max-size = <STM32MP_OPTEE_SIZE>;
 			id = <BL32_IMAGE_ID>;
 		};
+#else
+		tos_fw {
+			load-address = <0x0 DDR_SEC_BASE>;
+			max-size = <DDR_SEC_SIZE>;
+			id = <BL32_IMAGE_ID>;
+		};
+#endif /* STM32MP1_OPTEE_IN_SYSRAM */
 #else
 		tos_fw {
 			load-address = <0x0 STM32MP_BL32_BASE>;
diff --git a/plat/st/stm32mp1/platform.mk b/plat/st/stm32mp1/platform.mk
index abdf70896..0ac0f9590 100644
--- a/plat/st/stm32mp1/platform.mk
+++ b/plat/st/stm32mp1/platform.mk
@@ -50,6 +50,9 @@ endif
 # STM32 image header version v2.0
 STM32_HEADER_VERSION_MAJOR:=	2
 STM32_HEADER_VERSION_MINOR:=	0
+
+# OP-TEE cannot be in SYSRAM on STM32MP13
+override STM32MP1_OPTEE_IN_SYSRAM :=	0
 endif
 
 ifeq ($(STM32MP15),1)
@@ -62,6 +65,8 @@ STM32_HEADER_VERSION_MAJOR:=	1
 STM32_HEADER_VERSION_MINOR:=	0
 STM32MP_CRYPTO_ROM_LIB :=	1
 
+STM32MP1_OPTEE_IN_SYSRAM ?=	0
+
 # Decryption support
 ifneq ($(DECRYPTION_SUPPORT),none)
 $(error "DECRYPTION_SUPPORT not supported on STM32MP15")
@@ -166,6 +171,7 @@ $(eval $(call assert_booleans,\
 		STM32MP_USE_EXTERNAL_HEAP \
 		STM32MP13 \
 		STM32MP15 \
+		STM32MP1_OPTEE_IN_SYSRAM \
 )))
 
 $(eval $(call assert_numerics,\
@@ -194,6 +200,7 @@ $(eval $(call add_defines,\
 		STM32MP_USE_EXTERNAL_HEAP \
 		STM32MP13 \
 		STM32MP15 \
+		STM32MP1_OPTEE_IN_SYSRAM \
 )))
 
 # Include paths and source files
diff --git a/plat/st/stm32mp1/stm32mp1_fip_def.h b/plat/st/stm32mp1/stm32mp1_fip_def.h
index a71af5c6c..108680b52 100644
--- a/plat/st/stm32mp1/stm32mp1_fip_def.h
+++ b/plat/st/stm32mp1/stm32mp1_fip_def.h
@@ -92,7 +92,7 @@
 #define STM32MP_DTB_BASE		STM32MP_BL32_DTB_BASE
 #endif
 
-#ifdef AARCH32_SP_OPTEE
+#if defined(AARCH32_SP_OPTEE) && STM32MP1_OPTEE_IN_SYSRAM
 #define STM32MP_OPTEE_BASE		STM32MP_SEC_SYSRAM_BASE
 
 #define STM32MP_OPTEE_SIZE		(STM32MP_BL2_DTB_BASE -  \
-- 
2.47.2

